examples/tree/plot_unveil_tree_structure.py

```diff
@@ -16,119 +16,169 @@
-# tree structure and allows access to low level attributes. The binary tree
-# tree_ is represented as a number of parallel arrays. The i-th element of each
-# array holds information about the node `i`. Node 0 is the tree's root. NOTE:
-# Some of the arrays only apply to either leaves or split nodes, resp. In this
-# case the values of nodes of the other type are arbitrary!
+##############################################################################
+# Tree structure
+# --------------
 #
-# Among those arrays, we have:
-#   - left_child, id of the left child of the node
-#   - right_child, id of the right child of the node
-#   - feature, feature used for splitting the node
-#   - threshold, threshold value at the node
+# The decision classifier has an attribute called ``tree_`` which allows access
+# to low level attributes such as ``node_count``, the total number of nodes,
+# and ``max_depth``, the maximal depth of the tree. It also stores the
+# entire binary tree structure, represented as a number of parallel arrays. The
+# i-th element of each array holds information about the node ``i``. Node 0 is
+# the tree's root. Some of the arrays only apply to either leaves or split
+# nodes. In this case the values of the nodes of the other type is arbitrary.
+# For example, the arrays ``feature`` and ``threshold`` only apply to split
+# nodes. The values for leaf nodes in these arrays are therefore arbitrary.
 #
+# Among these arrays, we have:
+#
+#   - ``children_left[i]``: id of the left child of node ``i`` or -1 if leaf
+#     node
+#   - ``children_right[i]``: id of the right child of node ``i`` or -1 if leaf
+#     node
+#   - ``feature[i]``: feature used for splitting node ``i``
+#   - ``threshold[i]``: threshold value at node ``i``
+#   - ``n_node_samples[i]``: the number of of training samples reaching node
+#     ``i``
+#   - ``impurity[i]``: the impurity at node ``i``
+#
+# Using the arrays, we can traverse the tree structure to compute various
+# properties. Below, we will compute the depth of each node and whether or not
+# it is a leaf.
 
-# Using those arrays, we can parse the tree structure:
-
-n_nodes = estimator.tree_.node_count
-children_left = estimator.tree_.children_left
-children_right = estimator.tree_.children_right
-feature = estimator.tree_.feature
-threshold = estimator.tree_.threshold
-
+n_nodes = clf.tree_.node_count
+children_left = clf.tree_.children_left
+children_right = clf.tree_.children_right
+feature = clf.tree_.feature
+threshold = clf.tree_.threshold
 
-# The tree structure can be traversed to compute various properties such
-# as the depth of each node and whether or not it is a leaf.
 node_depth = np.zeros(shape=n_nodes, dtype=np.int64)
 is_leaves = np.zeros(shape=n_nodes, dtype=bool)
-stack = [(0, -1)]  # seed is the root node id and its parent depth
+stack = [(0, -1)]  # start with the root node id (0) and its parent depth (-1)
```

Comment:
kind of a nit but this could just be [(0, 0)]  and we could have

`node_id, current_depth = stack.pop()`

below. It seems unnecessarily contrived to deal with the parent's depth

Reply:
You're right, I don't know why I didn't question this.